aws/kinesis/core/retrier.cc

```diff
@@ -192,37 +194,51 @@ void Retrier::fail(const std::shared_ptr<UserRecord>& ur,
 bool Retrier::succeed_if_correct_shard(const std::shared_ptr<UserRecord>& ur,
                                        TimePoint start,
                                        TimePoint end,
-                                       const std::string& shard_id,
+                                       const std::string& shard_id,                                     
                                        const std::string& sequence_number,
-                                       const bool should_invalidate_on_incorrect_shard) {
+                                       const bool should_invalidate_on_incorrect_shard,
+                                       const boost::optional<std::pair<uint128_t, uint128_t>>& hashrange_actual_shard) {
   const uint64_t actual_shard = ShardMap::shard_id_from_str(shard_id);
-  if (ur->predicted_shard() &&
-      *ur->predicted_shard() != actual_shard) {
-    //We should call invalidate only if:
-    // 1. If we are told to invalidate on incorrect shard.
-    if (should_invalidate_on_incorrect_shard) {
-      LOG(warning) << "Record went to shard " << shard_id << " instead of the "
+  if (ur->predicted_shard() && *ur->predicted_shard() != actual_shard) {
+    // retry if shard is not found or hashrange of the user record doesn't fit into the actual shard's hashrange
+    if (!hashrange_actual_shard || !((*hashrange_actual_shard).first <= ur->hash_key() && 
+        (*hashrange_actual_shard).second >= ur->hash_key())) {
+        // invalidate because this is a new shard or shard felt outside of actual shards hashrange.
+        invalidate_cache(ur, start, actual_shard, should_invalidate_on_incorrect_shard);
+
+        retry_not_expired(ur,
+                          start,
+                          end,
+                          "Wrong Shard",
+                          "Record " + std::to_string(ur->source_id()) + " did not end up in expected shard.");
+        return false;
+    } 
+    // child shard is numbered higher than the ancestor. if we landed on a child shard it means the parent shard can
+    // be removed now from the in-memory map. This will help the shardmap to converge
+    else if (*ur->predicted_shard() < actual_shard) {
+      invalidate_cache(ur, start, actual_shard, should_invalidate_on_incorrect_shard);
+    }
+  }
+  finish_user_record(
+      ur,
+      Attempt()
+          .set_start(start)
+          .set_end(end)
+          .set_result(shard_id, sequence_number));
+  return true;
+}
+
+void Retrier::invalidate_cache(const std::shared_ptr<UserRecord>& ur,
+                               const TimePoint start,
+                               const uint64_t& actual_shard,
+                               bool should_invalidate_on_incorrect_shard) {
+ if (should_invalidate_on_incorrect_shard) {
```

Comment:
1. formatting (align if with closing bracket)
2. should_invalidate_on_incorrect_shard should be const bool